CIP-????/README.md

```diff
@@ -0,0 +1,294 @@
+implementation would be called from Haskell using direct bindings to C. These
+bindings would be defined in `cardano-base`, using its existing `DSIGN`
+interface, with new builtins in Plutus Core on the basis of the `DSIGN`
+interface for both schemes.
+
+The builtins would be costed as follows: each signature verification has
+linear cost proportional to the number of message bits, and square with respect to key length.
+_Constants will be proportional to the ratio of time as benchmarked against
+Ed25519._
+
+Since we expect adding more cryptographic signature algorithms as they evolve,
+we should add datatype for selecting cryptographic algorithm:
+
+```
+data SignAlgo = Ed25519
+              | Blake2
+              | EcDSA
+              | SECP256k1
+              | Schnorr
+              | Falcon
+              | Sphincs
+              | Dilithium
+              | Sequoa
+```
+To facilitate extensibility, the name of algorithm should be rendered as case-insensitive text in JavaScript
+([CIP-0030](https://developers.cardano.org/docs/governance/cardano-improvement-proposals/cip-0030/)).
+
+| Cryptographic algorithm | Public key (bytes) | Signature (bytes) | Relative cost of verification (CPU) | Standardization |
+|:------------------------|-------------------:|------------------:|------------------------------------:|----------------:|
+| Ed25519                 |                 64 |                64 |                                   1 |    [Draft](https://www.nist.gov/news-events/news/2019/10/digital-signatures-and-elliptic-curve-cryptography-request-comments-draft) |
+| Falcon512               |                897 |               666 |                                 0.5 | [Selected](https://csrc.nist.gov/projects/post-quantum-cryptography/selected-algorithms-2022) | 
+| Dilithium2              |               1312 |              2420 |                                 0.5 | [Selected](https://csrc.nist.gov/projects/post-quantum-cryptography/selected-algorithms-2022) |
+| Sphincs+                |                 32 |              8080^[https://sofiaceli.com/2022/07/05/pq-signatures.html#:~:text=A%20table%20of-,comparison%20of%20sizes,-%3A] |                                 2.8^[Benchmark was for Sphins\*128s] | [Selected](https://csrc.nist.gov/projects/post-quantum-cryptography/selected-algorithms-2022) |
+| sequoa863pq128          |                108 |               216 |                                    ? | Submitted^[Communication on ebats@cryp.to mailing list.] |
+
+  : Reference for signature, key sizes, and time cost multipliers [as suggested here](https://blog.cloudflare.com/nist-post-quantum-surprise/). In the order of preference, with comparison against current default signature algorithm (Ed25519).
+
+The following function should then be used to verify arbitrary binary signed
+with permitted signature scheme:
+
+```
+verifySignWithAlgo :: SignAlgo
```

Comment:
I think this is the heart of the "extensibility" proposal. The obvious question is "why not just write this function in user-space, having it delegate to the appropriate single-algorithm builtin?"

Here's my guess at the scenario you're trying to support where that wouldn't be enough (please actually write it down yourself though, I'd prefer not to guess!):

1. Dapp creator writes script S using `verifySignWithAlgo`
2. S can dynamically (?) choose which signature algorithm it is using, maybe by doing something with a value in the datum (?)
3. We update the Plutus ledger language being used by S such that `verifySignWithAlgo` has additional algorithms (I'm actually not sure we can do this safely!)
4. S can be used as-is, only requiring a change to e.g. its datum to get it to use the new algorithm.

This has several tricky steps, notably 2 and 3. It's hard for me to say whether or not we could do that since I'm not sure what you're trying to do, but I think we might not be able to do those (esp 3) without e.g. making some scripts succeed that previously would have failed at runtime, which we shouldn't really do.